RecoPixelVertexing/PixelTriplets/plugins/CAHitQuadrupletGeneratorGPU.cu

```diff
@@ -0,0 +1,498 @@
+  cudaMemset(device_isOuterHitOfCell, 0,
+             maxNumberOfLayers * maxNumberOfHits *
+                 sizeof(GPUSimpleVector<maxCellsPerHit, unsigned int>));
+  h_foundNtuplets.resize(maxNumberOfRegions);
+  d_foundNtuplets.resize(maxNumberOfRegions);
+  for (int i = 0; i < maxNumberOfRegions; ++i) {
+    cudaMalloc(&d_foundNtuplets[i],
+               sizeof(GPUSimpleVector<maxNumberOfQuadruplets, Quadruplet>));
+    cudaMallocHost(&h_foundNtuplets[i],
+                   sizeof(GPUSimpleVector<maxNumberOfQuadruplets, Quadruplet>));
+  }
+
+  cudaMallocHost(&tmp_layers, maxNumberOfLayers * sizeof(GPULayerHits));
+  cudaMallocHost(&tmp_layerDoublets,
+                 maxNumberOfLayerPairs * sizeof(GPULayerDoublets));
+  cudaMallocHost(&h_layers, maxNumberOfLayers * sizeof(GPULayerHits));
+}
+
+void CAHitQuadrupletGeneratorGPU::launchKernels(const TrackingRegion &region,
+                                                int regionIndex) {
+
+  assert(regionIndex < maxNumberOfRegions);
+  dim3 numberOfBlocks_create(32, numberOfLayerPairs);
+  dim3 numberOfBlocks_connect(16, numberOfLayerPairs);
+  dim3 numberOfBlocks_find(8, numberOfRootLayerPairs);
+  ((GPUSimpleVector<maxNumberOfQuadruplets, Quadruplet>
+        *)(h_foundNtuplets[regionIndex]))
+      ->reset();
+  kernel_create<<<numberOfBlocks_create, 32, 0, cudaStream_>>>(
+      numberOfLayerPairs, d_doublets, d_layers, (GPUCACell *)device_theCells,
+      (GPUSimpleVector<maxCellsPerHit, unsigned int> *)device_isOuterHitOfCell,
+      (GPUSimpleVector<maxNumberOfQuadruplets, Quadruplet> *)
+          d_foundNtuplets[regionIndex],
+      region.origin().x(), region.origin().y(), maxNumberOfDoublets,
+      maxNumberOfHits);
+
+  kernel_connect<<<numberOfBlocks_connect, 512, 0, cudaStream_>>>(
+      numberOfLayerPairs, d_doublets, (GPUCACell *)device_theCells,
+      (GPUSimpleVector<maxCellsPerHit, unsigned int> *)device_isOuterHitOfCell,
+      region.ptMin(), region.origin().x(), region.origin().y(),
+      region.originRBound(), caThetaCut, caPhiCut, caHardPtCut,
+      maxNumberOfDoublets, maxNumberOfHits);
+
+  kernel_find_ntuplets<<<numberOfBlocks_find, 1024, 0, cudaStream_>>>(
+      numberOfRootLayerPairs, d_doublets, (GPUCACell *)device_theCells,
+      (GPUSimpleVector<maxNumberOfQuadruplets, Quadruplet> *)
+          d_foundNtuplets[regionIndex],
+      d_rootLayerPairs, 4, maxNumberOfDoublets);
+
+  cudaMemcpyAsync(h_foundNtuplets[regionIndex], d_foundNtuplets[regionIndex],
+                  sizeof(GPUSimpleVector<maxNumberOfQuadruplets, Quadruplet>),
+                  cudaMemcpyDeviceToHost, cudaStream_);
+}
+
+std::vector<std::array<std::pair<int, int>, 3>>
+CAHitQuadrupletGeneratorGPU::fetchKernelResult(int regionIndex) {
+
+  cudaMemsetAsync(device_isOuterHitOfCell, 0,
+                  maxNumberOfLayers * maxNumberOfHits *
+                      sizeof(GPUSimpleVector<maxCellsPerHit, unsigned int>),
+                  cudaStream_);
```

Comment:
Could you add a comment that this resets temporary memory for the next event, and is not needed for reading the output? Just to explain why it is ok to have async call here without `cudaStreamSynchronize()`.